pulp_2to3_migration/app/migration.py

```diff
@@ -320,6 +324,89 @@ def create_repo_version(progress_rv, pulp2_repo, pulp3_remote=None):
         pulp2_repo(Pulp2Repository): a pre-migrated repository to create a repo version for
         pulp3_remote(remote): a pulp3 remote
     """
+    def detect_path_overlap(paths):
+        """
+        Check for valid POSIX paths (ie ones that aren't duplicated and don't overlap).
+
+        Overlapping paths are where one path terminates inside another (e.g. a/b and a/b/c).
+
+        NOTE: The logic is copied from pulpcore.app.files.validate_file_paths().
+
+        This function returns the first dupe or overlap it detects. We use a trie (or
+        prefix tree) to keep track of which paths we've already seen.
+
+        Args:
+            paths (iterable of str): An iterable of strings each representing a relative path
+
+        Returns:
+            str: a path which overlaps or duplicates another
+
+        """
+        path_trie = StringTrie(separator="/")
+        for path in paths:
+            if path in path_trie:
+                # path duplicates a path already in the trie
+                return path
+
+            if path_trie.has_subtrie(path):
+                # overlap where path is 'a/b' and trie has 'a/b/c'
+                return path
+
+            prefixes = list(path_trie.prefixes(path))
+            if prefixes:
+                # overlap where path is 'a/b/c' and trie has 'a/b'
+                return path
+
+            # if there are no overlaps, add it to our trie and continue
+            path_trie[path] = True
+
+    def resolve_path_overlap(version):
+        """
+        Remove content which path overlaps some other.
+
+        If it's a duplicated path, remove the older content.
+
+        If something is absolutely wrong and we were not able to resolve conflicts,
+        repo version creation will fail later.
+
+        Paths can be overlapping because of an old Pulp 2 bug.
+
+        Args:
+            version(pulpcore.app.model.RepositoryVersion): incomplete version which needs path
+                overlap resolution
+
+        """
+        paths = ContentArtifact.objects.filter(content__pk__in=version.content).values_list(
+            "relative_path", flat=True
+        )
+        paths = list(paths)
+        max_conflicts = version.content.count() - 1
+        for i in range(max_conflicts):
+            bad_path = detect_path_overlap(paths)
+            if not bad_path:
+                # no path overlaps, we are good
+                break
```

Comment:
Should this be `continue`?  Do we need to keep checking other paths?

Reply:
Ah, maybe not.  Because we're passing the whole list in here.  The loop confused me, but I guess it's only being used for resolution of overlaps?

Reply:
I could have had a while loop here, but just to be on the safe side, in case things go south, we are not stuck in the infinite loop.
It will loop only if there are any conflicts.